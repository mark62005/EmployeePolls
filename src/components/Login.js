import PropTypes from "prop-types";
import { useState, useEffect } from "react";
import { connect } from "react-redux";
import { setAuthedUser } from "../actions/authedUser";
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import Alert from "react-bootstrap/Alert";

const Login = ({ dispatch, users, authedUser }) => {
    const [ userId, setUserId ] = useState("");
    const [ password, setPassword ] = useState("");
    const [ isValidated, setIsValidated ] = useState(true);
    const [ errorMessage, setErrorMessage ] = useState("");

    const handleInputChange = (e, field) => {
        e.preventDefault();

        if (field === "uid") {
            setUserId(e.target.value);
        } else {
            setPassword(e.target.value);
        }
    };

    const validateLoginInfo = (uid, password) => {
        const user = Object.values(users).filter((u) => u.id === uid);

        if (user.length !== 1 || user[ 0 ] === undefined) {
            setErrorMessage("Error. User ID doesn't exist.");
            setIsValidated(false);
            return false;
        } else if (user[ 0 ].password !== password) {
            setErrorMessage("Error. Password is not correct.");
            setIsValidated(false);
            return false;
        }

        setIsValidated(true);
        return true;
    };

    const handleSubmit = (e) => {
        const form = e.currentTarget;

        if (
            validateLoginInfo(userId, password) === true
            && form.checkValidity() === true
            && isValidated === true
        ) {
            e.preventDefault();
            dispatch(setAuthedUser(userId));
            localStorage.setItem("authedId", JSON.stringify(userId));
            console.log(localStorage.getItem("authedId"));
        } else {
            e.stopPropagation();
        }
        setUserId("");
        setPassword("");
    };

    return (
        <Container fluid className="text-center w-50">
            <h1>Login</h1>
            <Form noValidate >
                <Row className="text-start">
                    <Form.Group
                        as={ Col }
                        className="mb-3"
                        controlId="userId"
                    >
                        <Form.Label>ID</Form.Label>
                        <Form.Control
                            type="text"
                            placeholder="Enter Your ID"
                            value={ userId }
                            onChange={ (e) => handleInputChange(e, "uid") }
                            required
                        />
                        <Form.Control.Feedback type="invalid">
                            Please provide a valid User ID.
                        </Form.Control.Feedback>
                    </Form.Group>
                </Row>
                <Row className="text-start">
                    <Form.Group as={ Col } className="mb-3" controlId="password">
                        <Form.Label>Password</Form.Label>
                        <Form.Control
                            type="password"
                            placeholder="Enter Your Password"
                            value={ password }
                            onChange={ (e) => handleInputChange(e, "password") }
                            required
                        />
                        <Form.Control.Feedback type="invalid">
                            Please provide a valid password.
                        </Form.Control.Feedback>
                    </Form.Group>
                </Row>
                {
                    isValidated === true ? null : <Alert variant="danger"> { errorMessage } </Alert>
                }
                <Button
                    variant="primary"
                    onClick={ handleSubmit }
                    disabled={ userId === "" || password === "" }
                >
                    Submit
                </Button>
            </Form>
        </Container>
    );
};

Login.propTypes = {
    dispatch: PropTypes.func.isRequired,
    users: PropTypes.object.isRequired,
};

const mapStateToProps = ({ users, authedUser }) => ({
    users,
    authedUser,
});

export default connect(mapStateToProps)(Login);