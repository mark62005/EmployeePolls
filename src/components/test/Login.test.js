import {
    fireEvent,
    render,
    screen,
} from "@testing-library/react";
import thunk from "redux-thunk";
import reducer from "../../reducers";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import Login from "../Login";

describe("Login", () => {
    var mockStore;

    beforeEach(() => {
        mockStore = configureStore({ reducer, middleware: [ thunk ] });
    });

    it("will match the snapshot", () => {
        var view = render(
            <MemoryRouter>
                <Provider store={ mockStore }>
                    <Login />
                </Provider>
            </MemoryRouter>
        );

        expect(view).toMatchSnapshot();
    });

    it("will display a login form with user id select and a sign in button", async () => {
        render(
            <MemoryRouter>
                <Provider store={ mockStore }>
                    <Login />
                </Provider>
            </MemoryRouter>
        );

        var idSelect = screen.getByTestId("user-id-select");
        var signInButton = screen.getByTestId("sign-in-button");

        expect(idSelect).toBeInTheDocument();
        expect(signInButton).toBeInTheDocument();
    });

    it("will not display the login form if sign in is successful", async () => {
        var view = render(
            <MemoryRouter>
                <Provider store={ mockStore }>
                    <Login />
                </Provider>
            </MemoryRouter>
        );

        var idSelect = screen.getByTestId("user-id-select");
        fireEvent.change(idSelect, { target: { value: "sarahedo" } });
        var signInButton = screen.getByTestId("sign-in-button");
        fireEvent.click(signInButton);

        view.rerender();
        expect(idSelect).not.toBeInTheDocument();
        expect(signInButton).not.toBeInTheDocument();
    });

    // it("will display a login form with id input, passsword input and a sign in button", async () => {
    //     render(
    //         <MemoryRouter>
    //             <Provider store={ mockStore }>
    //                 <Login />
    //             </Provider>
    //         </MemoryRouter>
    //     );

    //     var idInput = screen.getByTestId("id-input");
    //     var passwordInput = screen.getByTestId("password-input");
    //     var signInButton = screen.getByTestId("sign-in-button");
    //     var errorAlert = screen.queryByTestId("error-message");

    //     expect(idInput).toBeInTheDocument();
    //     expect(passwordInput).toBeInTheDocument();
    //     expect(signInButton).toBeInTheDocument();
    //     expect(errorAlert).toBeNull();
    // });

    // it("will display an error message if user id does not exist", async () => {
    //     render(
    //         <MemoryRouter>
    //             <Provider store={ mockStore }>
    //                 <Login />
    //             </Provider>
    //         </MemoryRouter>
    //     );

    //     var idInput = screen.getByTestId("id-input");
    //     fireEvent.change(idInput, { target: { value: "id" } });

    //     var passwordInput = screen.getByTestId("password-input");
    //     fireEvent.change(passwordInput, { target: { value: "password" } });

    //     var signInButton = screen.getByTestId("sign-in-button");
    //     fireEvent.click(signInButton);

    //     var errorAlert = screen.getByText("Error. User ID doesn't exist.");

    //     expect(idInput).toBeInTheDocument();
    //     expect(passwordInput).toBeInTheDocument();
    //     expect(signInButton).toBeInTheDocument();
    //     expect(errorAlert).toBeInTheDocument();
    // });
});