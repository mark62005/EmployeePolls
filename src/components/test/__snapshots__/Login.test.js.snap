// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login will match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="d-flex flex-column align-items-center justify-content-center container"
        style="height: 100vh; width: 100vw;"
      >
        <div
          class="text-center w-50 container-fluid"
        >
          <h1>
            Login
          </h1>
          <select
            class="my-4 form-select"
            data-testid="user-id-select"
          />
          <button
            class="btn btn-primary"
            data-testid="sign-in-button"
            type="button"
          >
            SIGN IN
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="d-flex flex-column align-items-center justify-content-center container"
      style="height: 100vh; width: 100vw;"
    >
      <div
        class="text-center w-50 container-fluid"
      >
        <h1>
          Login
        </h1>
        <select
          class="my-4 form-select"
          data-testid="user-id-select"
        />
        <button
          class="btn btn-primary"
          data-testid="sign-in-button"
          type="button"
        >
          SIGN IN
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
